src/parser/shadowlands/modules/items/dungeons/ShadowgraspTotem.tsx

```diff
@@ -0,0 +1,205 @@
+import Events, { CastEvent, DamageEvent, HealEvent, Item } from 'parser/core/Events';
+import Abilities from 'parser/core/modules/Abilities';
+import CastEfficiency from 'parser/shared/modules/CastEfficiency';
+import SpellUsable from 'parser/shared/modules/SpellUsable';
+import BoringItemValueText from 'parser/ui/BoringItemValueText';
+import ItemDamageDone from 'parser/ui/ItemDamageDone';
+import Statistic from 'parser/ui/Statistic';
+import STATISTIC_CATEGORY from 'parser/ui/STATISTIC_CATEGORY';
+import STATISTIC_ORDER from 'parser/ui/STATISTIC_ORDER';
+
+const ITEM: Item = ITEMS.SHADOWGRASP_TOTEM;
+const SUMMON_SPELL: Spell = SPELLS.SHADOWGRASP_TOTEM_SUMMON;
+const DAMAGE_SPELL: Spell = SPELLS.SHADOWGRASP_TOTEM_DAMAGE;
+const HEAL_SPELL: Spell = SPELLS.SHADOWGRASP_TOTEM_HEAL;
+const COOLDOWN_REDUCTION_SECONDS = 15;
+const DEFAULT_COOLDOWN_SECONDS = 60 * 2;
+const DEFAULT_TICK_RATE_MS = 1_000;
+
+function calculateHaste(baseTickRate: number, currentTickRate: number) {
+  return baseTickRate / currentTickRate - 1;
+}
+
+/**
+ * Track damage, healing and cooldown reset by Shadowgrasp Totem
+ *
+ * ## Example log:
+ *
+ * - https://www.warcraftlogs.com/reports/fZGa6X23CWbm9v8B#fight=9&type=damage-done&source=1
+ * - https://www.warcraftlogs.com/reports/twQhDxWCZN2jPbnp#fight=6&type=damage-done&source=13
+ */
+class ShadowgraspTotem extends Analyzer {
+  static dependencies = {
+    abilities: Abilities,
+    castEfficiency: CastEfficiency,
+    spellUsable: SpellUsable,
+  };
+  protected abilities!: Abilities;
+  protected castEfficiency!: CastEfficiency;
+  protected spellUsable!: SpellUsable;
+
+  private item: Item;
+
+  /** Timestamps of damage events since the last cast */
+  private currentDot?: number[];
+  /** All timestamps of all casts */
+  private dots: number[][] = [];
+  private totalDamage = 0;
+  private healing: number[] = [];
+
+  constructor(
+    options: Options & {
+      abilities: Abilities;
+      castEfficiency: CastEfficiency;
+      spellUsable: SpellUsable;
+    },
+  ) {
+    super(options);
+
+    this.item = this.selectedCombatant.getItem(ITEM.id)!;
+    if (this.item == null) {
+      this.active = false;
+      return;
+    }
+
+    // Add cast as an ability to show effective usage and cooldown in timeline
+    options.abilities.add({
+      spell: SUMMON_SPELL.id,
+      category: Abilities.SPELL_CATEGORIES.ITEMS,
+      cooldown: DEFAULT_COOLDOWN_SECONDS,
+      gcd: null,
+      castEfficiency: {
+        suggestion: true,
+        recommendedEfficiency: 0.8,
+      },
+    });
+
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SUMMON_SPELL), this.onCast);
+    this.addEventListener(Events.damage.by(SELECTED_PLAYER_PET).spell(DAMAGE_SPELL), this.onDamage);
+    this.addEventListener(Events.heal.to(SELECTED_PLAYER).spell(HEAL_SPELL), this.onHeal);
+  }
+
+  onCast(event: CastEvent) {
+    this.dots.push((this.currentDot = []));
```

Comment:
```suggestion
    this.currentDot = [];
    this.dots.push(this.currentDot);
```

don't use assignment as an expression, it is prone to errors